File: openstudio-hvac/test/test_cbecs_names.rb
require_relative 'test_helper'

# CBECS descriptive names mapped onto the gem's families (first increment: the
# clean-topology matches). Zone partitioning (heated-only vs cooled, unit heaters
# for leftovers) remains the caller's job, as in openstudio-standards cbecs_hvac.
class TestCbecsNames < Minitest::Test
  include FixtureHelper

  def test_baseboard_electric
    model = load_fixture
    zones = model.getThermalZones.sort_by(&:nameString)
    result = OpenStudioHVAC.build_system(model, 'Baseboard electric', zones)

    assert_empty result.air_loops, 'baseboards-only: no air system'
    assert_equal zones.size, model.getZoneHVACBaseboardConvectiveElectrics.size
    assert_empty model.getAirLoopHVACs
    assert_empty model.getPlantLoops
  end

  def test_baseboard_gas_boiler
    model = load_fixture
    zones = model.getThermalZones.sort_by(&:nameString)
    OpenStudioHVAC.build_system(model, 'Baseboard gas boiler', zones)

    assert_equal zones.size, model.getZoneHVACBaseboardConvectiveWaters.size
    assert_equal 2, model.getBoilerHotWaters.size
    assert(model.getBoilerHotWaters.all? { |b| b.fuelType == 'NaturalGas' })
  end

  def test_psz_ac_gas_coil_builds_one_unit_per_zone
    model = load_fixture
    zones = model.getThermalZones.sort_by(&:nameString)
    result = OpenStudioHVAC.build_system(model, 'PSZ-AC with gas coil', zones)

    # CBECS/90.1 convention: one packaged unit per zone
    assert_equal zones.size, result.air_loops.size
    assert(result.air_loops.all? { |al| al.thermalZones.size == 1 })
    assert_equal zones.size, model.getCoilCoolingDXSingleSpeeds.size
    assert_equal zones.size, model.getCoilHeatingGass.size
    assert_equal zones.size, model.getSetpointManagerSingleZoneReheats.size
    # each unit controls its own zone
    model.getSetpointManagerSingleZoneReheats.each do |spm|
      al = spm.airLoopHVAC.get
      assert_equal al.thermalZones.first.handle.to_s, spm.controlZone.get.handle.to_s
    end
    assert_empty model.getZoneHVACBaseboardConvectiveElectrics, "baseboard_type 'None'"
  end

  def test_psz_ac_electric_coil
    model = load_fixture
    zones = model.getThermalZones.sort_by(&:nameString)
    OpenStudioHVAC.build_system(model, 'PSZ-AC with electric coil', zones)
    assert_equal zones.size, model.getCoilHeatingElectrics.size
    assert_empty model.getCoilHeatingGass
  end

  def test_catalog_filter_by_origin
    cbecs = OpenStudioHVAC.systems.select { |r| r['origin'] == 'cbecs' }
    assert_operator cbecs.size, :>=, 4
  end
end

File: python/tests/modeling/test_cbecs_names.py
import unittest

import btap.modeling as modeling
from tests.support import load_fixture, needs_sdk


@needs_sdk
class TestCbecsNames(unittest.TestCase):
    """CBECS descriptive names mapped onto the gem's families (first increment: the
    clean-topology matches). Zone partitioning (heated-only vs cooled, unit heaters
    for leftovers) remains the caller's job, as in openstudio-standards cbecs_hvac."""

    def test_baseboard_electric(self):
        from btap._compat import sorted_by_name

        model = load_fixture()
        zones = sorted_by_name(model.getThermalZones())
        result = modeling.build_system(model, 'Baseboard electric', zones)

        self.assertEqual(0, len(result.air_loops), 'baseboards-only: no air system')
        self.assertEqual(len(zones), len(model.getZoneHVACBaseboardConvectiveElectrics()))
        self.assertEqual(0, len(model.getAirLoopHVACs()))
        self.assertEqual(0, len(model.getPlantLoops()))

    def test_baseboard_gas_boiler(self):
        from btap._compat import sorted_by_name

        model = load_fixture()
        zones = sorted_by_name(model.getThermalZones())
        modeling.build_system(model, 'Baseboard gas boiler', zones)

        self.assertEqual(len(zones), len(model.getZoneHVACBaseboardConvectiveWaters()))
        self.assertEqual(2, len(model.getBoilerHotWaters()))
        self.assertTrue(all(b.fuelType() == 'NaturalGas' for b in model.getBoilerHotWaters()))

    def test_psz_ac_gas_coil_builds_one_unit_per_zone(self):
        from btap._compat import sorted_by_name

        model = load_fixture()
        zones = sorted_by_name(model.getThermalZones())
        result = modeling.build_system(model, 'PSZ-AC with gas coil', zones)

        # CBECS/90.1 convention: one packaged unit per zone
        self.assertEqual(len(zones), len(result.air_loops))
        self.assertTrue(all(len(al.thermalZones()) == 1 for al in result.air_loops))
        self.assertEqual(len(zones), len(model.getCoilCoolingDXSingleSpeeds()))
        self.assertEqual(len(zones), len(model.getCoilHeatingGass()))
        self.assertEqual(len(zones), len(model.getSetpointManagerSingleZoneReheats()))
        # each unit controls its own zone
        for spm in model.getSetpointManagerSingleZoneReheats():
            al = spm.airLoopHVAC().get()
            self.assertEqual(str(al.thermalZones()[0].handle()),
                             str(spm.controlZone().get().handle()))
        self.assertEqual(0, len(model.getZoneHVACBaseboardConvectiveElectrics()),
                         "baseboard_type 'None'")

    def test_psz_ac_electric_coil(self):
        from btap._compat import sorted_by_name

        model = load_fixture()
        zones = sorted_by_name(model.getThermalZones())
        modeling.build_system(model, 'PSZ-AC with electric coil', zones)
        self.assertEqual(len(zones), len(model.getCoilHeatingElectrics()))
        self.assertEqual(0, len(model.getCoilHeatingGass()))

    def test_catalog_filter_by_origin(self):
        # Ruby Hash#[] yields nil for rows without an 'origin' key; .get mirrors that.
        cbecs = [r for r in modeling.systems() if r.get('origin') == 'cbecs']
        self.assertGreaterEqual(len(cbecs), 4)


if __name__ == '__main__':
    unittest.main()
